crates/block-producer/src/mempool/dependency_graph.rs

```diff
@@ -0,0 +1,826 @@
+    }
+
+    /// Removes a set of previously processed nodes from the graph.
+    ///
+    /// This is used to bound the size of the graph by removing nodes once they are no longer
+    /// required.
+    ///
+    /// # Errors
+    ///
+    /// Errors if
+    ///   - any node is unknown
+    ///   - any node is __not__ processed
+    ///   - any parent node would be left unpruned
+    ///
+    /// The last point implies that all parents of the given nodes must either be part of the set,
+    /// or already been pruned.
+    ///
+    /// This method is atomic.
+    pub fn prune_processed(&mut self, keys: BTreeSet<K>) -> Result<Vec<V>, GraphError<K>> {
+        let missing_nodes = keys
+            .iter()
+            .filter(|key| !self.vertices.contains_key(key))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_nodes.is_empty() {
+            return Err(GraphError::UnknownNodes(missing_nodes));
+        }
+
+        let unprocessed = keys.difference(&self.processed).cloned().collect::<BTreeSet<_>>();
+        if !unprocessed.is_empty() {
+            return Err(GraphError::UnprocessedNodes(unprocessed));
+        }
+
+        // No parent may be left dangling i.e. all parents must be part of this prune set.
+        let dangling = keys
+            .iter()
+            .flat_map(|key| self.parents.get(key))
+            .flatten()
+            .filter(|parent| !keys.contains(parent))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !dangling.is_empty() {
+            return Err(GraphError::DanglingNodes(dangling));
+        }
+
+        let mut pruned = Vec::with_capacity(keys.len());
+
+        for key in keys {
+            let value = self.vertices.remove(&key).expect("Checked in precondition");
+            pruned.push(value);
+            self.processed.remove(&key);
+            self.parents.remove(&key);
+
+            let children = self.children.remove(&key).unwrap_or_default();
+
+            // Remove edges from children to this node.
+            for child in children {
+                if let Some(child) = self.parents.get_mut(&child) {
+                    child.remove(&key);
+                }
+            }
+        }
+
+        Ok(pruned)
+    }
+
+    /// Removes the set of nodes __and all descendents__ from the graph, returning all removed
+    /// values.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if any of the given nodes does not exist.
+    ///
+    /// This method is atomic.
+    pub fn purge_subgraphs(&mut self, keys: BTreeSet<K>) -> Result<Vec<V>, GraphError<K>> {
+        let missing_nodes = keys
+            .iter()
+            .filter(|key| !self.vertices.contains_key(key))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_nodes.is_empty() {
+            return Err(GraphError::UnknownNodes(missing_nodes));
+        }
+
+        let mut visited = keys.clone();
+        let mut to_process = keys;
```

Comment:
nit: would renaming `to_process` into `to_remove` be a bit more self-explanatory?